mne/time_frequency/csd.py

```diff
@@ -1,56 +1,512 @@
+                     deprecated)
+from ..viz.misc import plot_csd
 from ..time_frequency.multitaper import (dpss_windows, _mt_spectra,
                                          _csd_from_mt, _psd_from_mt_adaptive)
-
-from ..externals.six.moves import xrange as range
+from ..parallel import parallel_func
+from ..externals.h5io import read_hdf5, write_hdf5
 
 
 class CrossSpectralDensity(object):
     """Cross-spectral density.
 
+    Given a list of time series, the CSD matrix denotes for each pair of time
+    series, the cross-spectral density. This matrix is symmetric and internally
+    stored as a vector.
+
+    This object can store multiple CSD matrices: one for each frequency.
+    Use ``.get_data(freq)`` to obtain an CSD matrix as an ndarray.
+
     Parameters
     ----------
-    data : array of shape (n_channels, n_channels)
-        The cross-spectral density matrix.
+    data : ndarray, shape ((n_channels**2 + n_channels) / 2, n_frequencies)
+        For each frequency, the cross-spectral density matrix in vector format.
     ch_names : list of string
-        List of channels' names.
-    projs :
-        List of projectors used in CSD calculation.
-    bads :
-        List of bad channels.
-    freqs : float | list of float
-        Frequency or frequencies for which the CSD matrix was calculated. If a
-        list is passed, data is a sum across CSD matrices for all frequencies.
+        List of string names for each channel.
+    tmin : float
+        Start of the time window for which CSD was calculated in seconds.
+    tmax : float
+        End of the time window for which CSD was calculated in seconds.
+    frequencies : float | list of float | list of list of float
+        Frequency or frequencies for which the CSD matrix was calculated. When
+        averaging across frequencies (see the :func:`CrossSpectralDensity.mean`
+        function), this will be a list of lists that contains for each
+        frequency bin, the frequencies that were averaged. Frequencies should
+        always be sorted.
     n_fft : int
-        Length of the FFT used when calculating the CSD matrix.
+        The number of FFT points or samples that have been used in the
+        computation of this CSD.
+    projs : list of Projection | None
+        List of projectors to apply to timeseries data when using this CSD
+        object to compute a DICS beamformer. Defaults to ``None``, which means
+        no projectors will be applied.
+
+    See Also
+    --------
+    csd_epochs
+    csd_array
     """
 
-    def __init__(self, data, ch_names, projs, bads, freqs,
-                 n_fft):  # noqa: D102
-        self.data = data
-        self.dim = len(data)
-        self.ch_names = cp.deepcopy(ch_names)
-        self.projs = cp.deepcopy(projs)
-        self.bads = cp.deepcopy(bads)
-        self.freqs = np.atleast_1d(np.copy(freqs))
+    def __init__(self, data, ch_names, tmin, tmax, frequencies, n_fft,
```

Comment:
why is tmin, tmax necessary?

Reply:
Just for bookkeeping. I figured it would be useful to include in time window the `__repr__` string. I don't think many people would go around creating their own CSD objects, so having these extra parameters in the constructor should not be that bothersome. It's an easy change if you disagree.

Reply:
ok I would at least make tmin, tmax at the end of the list before projs and allow it to be None. If you compute CSD on a block on raw data then you don't have tmin, tmax

Reply:
ok!